### PR TITLE
Display property types in optional chains without optional type marker

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/37879

### CODE REVIEW

File: src/compiler/checker.ts
Comment (Andarist) [inline src/compiler/checker.ts:27530]: Note that this might not contain some guard rails related to control flow analysis that are present in `getTypeOfExpression`. I'm not sure if they are relevant for this case though.

Comment (rbuckton) [reply src/compiler/checker.ts:27530]: Wouldn't this work as well?

```ts
const type = removeOptionalTypeMarker(getTypeOfExpression(location as Expression));
```

For an expression like `a?.b.c`, hovering over `c` would still show `C | undefined` (since its the end of the chain and the optional type marker is replaced with `undefined`), but hovering over `b` would just show `B`.

The upside is that `getTypeOfExpression` caches, so you avoid recomputing a type that's already cached.

Comment (rbuckton) [reply src/compiler/checker.ts:27530]: There's no need to make this conditional on `location.flags & NodeFlags.OptionalChain`. The marker type will only be present for optional chains.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: src/compiler/checker.ts
Comment (Andarist) [reply src/compiler/checker.ts:27530]: cool, removed this redundant check 👍

**Comment by Andarist** [comment]
@sandersn hmm, it feels that the automation didn't move this to the "waiting on reviewers" column

**Comment by sandersn** [comment]
@Andarist yep, that part isn't automated yet

**Comment by Andarist** [comment]
@rbuckton @gabritto friendly 🏓

File: tests/cases/fourslash/quickInfoInOptionalChain.ts
Comment (jakebailey) [inline tests/cases/fourslash/quickInfoInOptionalChain.ts:25]: I'm confused as to the end result here; both `bar` and `baz` themselves are not declared as potentially undefined, and yet `bar` is not `| undefined` while `baz` is. Why should they behave differently?

Comment (Andarist) [reply tests/cases/fourslash/quickInfoInOptionalChain.ts:25]: I admit this is a little bit confusing and I'm open for feedback how to improve this further. 

`baz` has `| undefined` because it's the end of the optional chain expression and the whole optional chain expression has `| undefined`. Without this bit you'd look at an if statement that would always~ be truthy. That `| undefined` is here to inform you that the whole expression can indeed result in both branches to be taken.

OTOH, `bar` doesn't have it because with it it looks really weird that we can just chain off this expression (`foo?.bar`) without guarding the `.baz` property access

**Review by rbuckton — APPROVED** [review]
This looks good, though I'll give @jakebailey time to reply with any additional feedback before merging.

File: tests/cases/fourslash/quickInfoInOptionalChain.ts
Comment (jakebailey) [reply tests/cases/fourslash/quickInfoInOptionalChain.ts:25]: I guess that makes sense, though I'd probably personally defer to @DanielRosenwasser 😄

Comment (Andarist) [reply tests/cases/fourslash/quickInfoInOptionalChain.ts:25]: That's a good catch. I think that it's worth unifying those - I'd expect `x?.method;` to include `undefined` based on the given described intuition. I see how this is quite subtle though so perhaps not everybody would agree with me on this one and that's fine.

I think that this find doesn't have to block this PR, I can just prepare a follow up fixing it at some later point in time. Up to you though.

File: src/compiler/checker.ts
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:27612]: I'm not that familiar with this part of the code. Why doesn't this also apply to the entirety of `a?.b.c`?

Comment (rbuckton) [reply src/compiler/checker.ts:27612]: The "optional type" marker is not added at the outermost `PropertyAccessExpression`, only for inner optional chain parts. At the outermost `PropertyAccessExpression`, we just union it with the normal `undefined` type.

Comment (rbuckton) [reply src/compiler/checker.ts:27612]: See `propagateOptionalTypeMarker` in checker.ts for the logic related to when the marker type is added vs a regular `undefined`.

**Comment by rbuckton** [comment]
@DanielRosenwasser, @jakebailey any further feedback? If not, I will merge this.

**Comment by DanielRosenwasser** [comment]
I agree that my point doesn't need to hold back the PR

**Review by jakebailey — APPROVED** [review]

**Pull request merged**
